lib/chcol/variant.go

```diff
@@ -0,0 +1,126 @@
+// Licensed to ClickHouse, Inc. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. ClickHouse, Inc. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package chcol
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+)
+
+// Variant represents a ClickHouse Variant type that can hold multiple possible types
+type Variant struct {
+	value any
+}
+
+// NewVariant creates a new Variant with the given value
+func NewVariant(v any) Variant {
+	return Variant{value: v}
+}
+
+// Nil returns true if the underlying value is nil.
+func (v Variant) Nil() bool {
+	return v.value == nil
+}
+
+// Any returns the underlying value as any. Same as Interface.
+func (v Variant) Any() any {
+	return v.value
+}
+
+// Interface returns the underlying value as interface{}. Same as Any.
+func (v Variant) Interface() interface{} {
+	return v.value
+}
+
+// Int returns the value as an int if possible
+func (v Variant) Int() (int, bool) {
+	if i, ok := v.value.(int); ok {
+		return i, true
+	}
+
+	return 0, false
+}
+
+// Int64 returns the value as an int64 if possible
+func (v Variant) Int64() (int64, bool) {
+	if i, ok := v.value.(int64); ok {
+		return i, true
+	}
+
+	return 0, false
+}
+
+// String returns the value as a string if possible
+func (v Variant) String() (string, bool) {
+	if s, ok := v.value.(string); ok {
+		return s, true
+	}
+
+	return "", false
+}
+
+// Bool returns the value as an bool if possible
+func (v Variant) Bool() (bool, bool) {
+	if b, ok := v.value.(bool); ok {
+		return b, true
+	}
+
+	return false, false
+}
```

Comment:
I just wonder if we need these functions at all. See:

```go
package main

import "fmt"

type Variant struct {
	value any
}

func (v *Variant) Value() any {
	return v.value
}

func main() {
	v := Variant{}
	i, ok := v.Value().(int)
	fmt.Println(i, ok)
}
```

We can just expose value and let the user freely type assert.

Reply:
More: If we make `value` public, we don't need any value access functions.

Reply:
I'm considering removing `Variant` in favor of `VariantWithType` as the default. There are many cases where the ClickHouse type needs to be provided. If `value` were public then it doesn't make this type any different from a regular `any`.

For now I agree it may be best to simply make `value` public